lib/IDE/SourceEntityWalker.cpp

```diff
@@ -125,11 +125,15 @@ bool SemaAnnotator::walkToDeclPre(Decl *D) {
   bool IsExtension = false;
 
   if (auto *VD = dyn_cast<ValueDecl>(D)) {
-    if (VD->hasName() && !VD->isImplicit()) {
+    if (!VD->isImplicit()) {
       SourceManager &SM = VD->getASTContext().SourceMgr;
-      NameLen = VD->getBaseName().userFacingName().size();
-      if (Loc.isValid() && SM.extractText({Loc, 1}) == "`")
-        NameLen += 2;
+      if (VD->hasName()) {
+        NameLen = VD->getBaseName().userFacingName().size();
+        if (Loc.isValid() && SM.extractText({Loc, 1}) == "`")
+          NameLen += 2;
+      } else if (Loc.isValid() && SM.extractText({Loc, 1}) == "_") {
```

Comment:
I think this doesn’t work if the identifier starts with an underscore. E.g. `let _foo = 2`. If it does work, I think it would be worth adding a test case for it.

I also don’t understand yet why this should be necessary. If the identifier is empty, then `DeclBaseName::userFacingName()` should return a `_`, which has a length of 1. 🤔 
https://github.com/apple/swift/blob/7ccc1b054a607881978b8efa0d7042d8bcd3d63f/include/swift/AST/Identifier.h#L339

Reply:
If we have something like `_x`, `hasName()` returns true, so it should work. The reason why this is necessary for `_` is because it doesn't go down that branch if the identifier is empty, because `hasName()` returns false.

Reply:
Perhaps we could drop the `hasName()`-check completely since `userFacingName()` returns the right thing in any case?

Reply:
One thing to also consider is invalid declarations like `let = 1` or `struct {}` (not relevant for your implementation but for other users of `SemaAnnotator`), which also don’t have a name and whose name isn’t `_` either.

Just an idea if the logic starts to get too complicated here: You could also check for the standard straightforward cases here (no backticks, name not empty) and if that doesn’t match, use `Lexer::getTokenAtLocation` to look the actual source code up. The downside is that `Lexer::getTokenAtLocation` needs to set up a `Lexer` instance, which is slower (although it’s not really slow either and I think we’re doing it in a variety of places).

Reply:
Hm yes, dropping `hasName()` also seems to cause `test/IDE/annotation.swift` to fail as the annotator now reports things like
```swift
var <Var>protocolProperty1</Var>: <iStruct@>Int</iStruct> { <Accessor>g</Accessor>et }
```
instead of
```swift
var <Var>protocolProperty1</Var>: <iStruct@>Int</iStruct> { get }
```

I think the standard straightforward cases as well as the backtick and `_` case should be handled correctly with the PR implementation right now, though using a lexer might be a bit cleaner than looking up the source code directly.